Documentation for getting started with android (#341)
mozilla/experimenter-docs: docs/getting-started/android-integration.md

---
id: getting-started-for-android-engineers
title: Getting Started for Android Engineers
slug: /getting-started-for-android-engineers
---

# Introduction

Nimbus is an experimentation platform from Mozilla.

This document shows you how to set up the Nimbus SDK with a new Android app. It assumes that your app is already using the Glean SDK and Android Components.

# Building with Nimbus

> Nimbus is distributed as part of the `android-components`, which in turn bundles Rust code as part of a what Mozilla's Application Services ["Megazord"](https://github.com/mozilla/application-services/blob/main/docs/design/megazords.md).

Reviewer: ```suggestion
Nimbus is distributed as part of the `android-components`, which in turn bundles Rust code as part of Mozilla's Application Services ["Megazord"](https://github.com/mozilla/application-services/blob/main/docs/design/megazords.md).
```

    }
}
```

and in `app/build.gradle`:

```gradle
apply plugin: "org.mozilla.components.nimbus-gradle-plugin"

nimbus {
    // The path to the Nimbus feature manifest file
    manifestFile = "nimbus.fml.yaml"

    // Map from the variant name to the channel as experimenter and nimbus understand it.
    // If nimbus's channels were accurately set up well for this project, then this
    // shouldn't be needed.
    channels = [
            debug: "debug",
            nightly: "nightly",
            beta: "beta",
            release: "release",
    ]
    // This is generated by the FML and should be checked into git.
    // It will be fetched by Experimenter (the Nimbus experiment website)
    // and used to inform experiment configuration.
    experimenterManifest = ".experimenter.yaml"
}
```

In this case, it should generate a file named in the `nimbus.fml.yaml` file. In the case of Fenix, this is called `FxNimbus`.

# The start-up sequence

Before using Nimbus in your Android app, you need to start it.

The Nimbus SDK is a configuration store, making configuration available to the any thread, and — to a first approximation— to be immutable within the same session of the app.

> For this reason, we want to be start the Nimbus SDK as close to the beginning of the start of the app as possible.

Reviewer: ```suggestion
For this reason, we want to be starting the Nimbus SDK as close to the beginning of the start of the app as possible.
```

    fun beginNimbusSetup() {
        Megazord.init()

        nimbus = createNimbus(this, NIMBUS_REMOTE_SETTINGS_ENDPOINT)
    }

    fun finishNimbusSetup() {
        nimbus.fetchExperiments()
    }

    fun createNimbus(context: Context, urlString: String): NimbusApi {
        val isAppFirstRun = context.settings().isFirstRun
        val customTargetingAttibutes = JSONObject().apply {
            // Put any custom attributes you want to use to segment an audience on to
            // target your experiments.
            put("is_first_run", isAppFirstRun)
        }

        val appInfo = NimbusAppInfo(
            appName = "my-app-name",
            channel = BuildConfig.BUILD_TYPE,
            customTargetingAttributes = customTargetingAttributes
        )

        // Use the Nimbus builder to build a NimbusApi object.
        return NimbusBuilder(context).apply {
            url = urlString
            errorReporter = { message, e ->
                Logger.error("Nimbus error: $message", e)
            }

        }.build(appInfo)
    }
}
```

Notes:

1. `Megazord.init()` is called before `createNimbus()`.
1. `createNimbus` uses a `NimbusBuilder` to create the Nimbus object.

> 1. Building a `JSONObject` of custom targeting attributes.

Reviewer: I think the wording of this might be more clear by using the same verb tense as the second bullet
```suggestion
1. We build a `JSONObject` of custom targeting attributes.
```

            }
            // …
        }.build(appInfo)
```

### Connecting the `NimbusApi` to FML generated code

The FML generated code has a runtime dependency on the `NimbusApi`.

Two callbacks can be provided to the `NimbusBuilder`, one to connect the `NimbusApi` object when it is created,
and one to clear the caching in the generated code.

```kotlin
        // Use the Nimbus builder to build a NimbusApi object.
        return NimbusBuilder(context).apply {
            // …
            // Connect FxNimbus to the Nimbus SDK.
            onCreateCallback = { nimbus ->
                FxNimbus.initialize { nimbus }
            }
            onApplyCallback = {
                FxNimbus.invalidateCachedValues()
            }
            // …
        }.build(appInfo)
```

Note:

`onApplyCallback` is a syntactic sugar for

```kotlin
nimbus.register(object : NimbusInterface.Observer {
    // do after every applyPendingExperiments()
    onApplyCallback(nimbus)
})
```

However, `onApplyCallback` in `NimbusBuilder` is guaranteed to be executed on the first `applyPendingExperiments()`, i.e. on every startup.

### Handling First Run experiments

> Since `fetchExperiments` from the remote settings URL is slow, and we wish to be able have access to the Nimbus experimental configuration as early in start up as possible, Nimbus downloads and caches the experiment recipes on the `n`th run of the app and only applies them and makes them available to the app at the beginning of the _next_ i.e. the `(n + 1)`th run of ther app.

Reviewer: ```suggestion
Since `fetchExperiments` from the remote settings URL is slow, and we wish to be able have access to the Nimbus experimental configuration as early in start up as possible, Nimbus downloads and caches the experiment recipes on the `n`th run of the app and only applies them and makes them available to the app at the beginning of the _next_ i.e. the `(n + 1)`th run of the app.
```

Astute readers will notice that when `n = 0`, i.e. the very first time the app is run, there are no experiment recipes downloaded. If Remote Settings experiment recipes JSON payload is available as a `raw/` resource, it can be loaded in at first run:

```kotlin
        // Use the Nimbus builder to build a NimbusApi object.
        return NimbusBuilder(context).apply {
            // …
            isFirstRun = isAppFirstRun
            initialExperiments = R.raw.initial_experiments
            timeoutLoadingExperiment = TIME_OUT_LOADING_EXPERIMENT_FROM_DISK_MS // defaults to 200 (ms)
            // …
        }.build(appInfo)
```

> The `iniital_experiments.json` file can be downloaded, either as part of the build, or in an automated/timed job. e.g. this is the [Github Action workflow used by Fenix](https://github.com/mozilla-mobile/fenix/blob/main/.github/workflows/fenix-update-nimbus-experiments.yml).

Reviewer: ```suggestion
The `initial_experiments.json` file can be downloaded, either as part of the build, or in an automated/timed job. e.g. this is the [Github Action workflow used by Fenix](https://github.com/mozilla-mobile/fenix/blob/main/.github/workflows/fenix-update-nimbus-experiments.yml).
```
